src/backends/gpu/multi_event_stream.cu

```diff
@@ -0,0 +1,156 @@
+#include <common_types.hpp>
+#include <backends/event.hpp>
+#include <backends/gpu/multi_event_stream.hpp>
+#include <memory/array.hpp>
+#include <memory/copy.hpp>
+#include <util/rangeutil.hpp>
+
+namespace nest {
+namespace mc {
+namespace gpu {
+
+namespace kernel {
+    template <typename T, typename I>
+    __global__ void mark_until_after(
+        I n,
+        I* mark,
+        const I* span_end,
+        const T* ev_time,
+        const T* t_until)
+    {
+        I i = threadIdx.x+blockIdx.x*blockDim.x;
+        if (i>=n) return;
+
+        auto t = t_until[i];
+        auto end = span_end[i];
+        auto &m = mark[i];
+
+        while (m!=end && !(ev_time[m]>t)) { ++m; }
+    }
+
+    template <typename T, typename I>
+    __global__ void drop_marked_events(
+        I n,
+        I* n_nonempty,
+        I* span_begin,
+        const I* span_end,
+        const I* mark)
+    {
+        I i = threadIdx.x+blockIdx.x*blockDim.x;
+        if (i>=n) return;
+
+        bool emptied = (span_begin[i]<span_end[i] && mark[i]==span_end[i]);
+        span_begin[i] = mark[i];
+        if (emptied) {
+            atomicAdd(n_nonempty, (cell_size_type)-1);
+        }
+    }
+
+    template <typename T, typename I>
+    __global__ void event_time_if_before(
+        I n,
+        const I* span_begin,
+        const I* span_end,
+        const T* ev_time,
+        T* t_until)
+    {
+        I i = threadIdx.x+blockIdx.x*blockDim.x;
+        if (i>=n) return;
+
+        if (span_begin[i]<span_end[i]) {
+            auto ev_t = ev_time[span_begin[i]];
+            if (t_until[i]>ev_t) {
+                t_until[i] = ev_t;
+            }
+        }
+    }
+} // namespace kernel
+
+void multi_event_stream::clear() {
+    memory::fill(span_begin_, 0u);
+    memory::fill(span_end_, 0u);
+    memory::fill(mark_, 0u);
+    n_nonempty_stream_[0] = 0;
+}
+
+void multi_event_stream::init(const std::vector<deliverable_event>& staged) {
```

Comment:
This was kept for consistency in interface with the other (cpu) implementation. I will change both to pass-by-value and move at the caller's end.